File: src/__tests__/enroll-sas.test.ts
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";
import {
	type EnrollParty,
	type EnrollRole,
	enrollCommitment,
	enrollCommitmentPreimage,
	enrollSas,
	enrollSasPreimage,
	verifyEnrollCommitment,
} from "../shared/cross-domain-sas.js";

////////////////////////////////
//  Owner-anchored role-tagged enroll SAS vectors
//
//  vectors.json is read by BOTH this suite and (in the Android pass) SasCryptoTest.kt, so
//  the hand-authored Kotlin twin cannot drift from this TS source. Unlike the gateway SAS,
//  the enroll preimage is FIXED-SLOT / role-tagged: it is NOT order-independent, and a
//  field-role swap changes the code (the injective property the red-team required).

interface EnrollCase {
	admin: EnrollParty;
	enrollee: EnrollParty;
	pin: string;
	sas: string;
}

const vectors = JSON.parse(
	fs.readFileSync(path.join(__dirname, "../../tests/fixtures/enroll-sas/vectors.json"), "utf8"),
) as {
	cases: EnrollCase[];
	roleSwap: EnrollCase & { differsFrom: string };
	fieldReassign: EnrollCase & { differsFrom: string };
	substitution: EnrollCase & { differsFrom: string };
	commitment: { party: EnrollParty; role: EnrollRole; salt: string; commitment: string };
	commitmentRoleBinding: {
		party: EnrollParty;
		role: EnrollRole;
		salt: string;
		commitment: string;
		differsFrom: string;
	};
	commitmentSaltBinding: {
		party: EnrollParty;
		role: EnrollRole;
		salt: string;
		commitment: string;
		differsFrom: string;
	};
	handComputed: EnrollCase & {
		role: EnrollRole;
		salt: string;
		sasPreimage: string;
		commitmentPreimage: string;
		commitment: string;
	};
};

describe("enroll SAS vectors", () => {
	it.each(vectors.cases.map((c, i) => [i, c] as const))("enrollSas reproduces case %i", (_, c) => {
		expect(enrollSas(c.admin, c.enrollee, c.pin)).toBe(c.sas);
		// Width is fixed at 6 digits so the two phones compare equal-length strings.
		expect(c.sas).toMatch(/^\d{6}$/);
	});

	it("is role-FIXED (swapping ADMIN and ENROLLEE yields a different code, NOT order-independent)", () => {
		const { admin, enrollee, pin, sas, differsFrom } = vectors.roleSwap;
		expect(enrollSas(admin, enrollee, pin)).toBe(sas);
		expect(sas).not.toBe(differsFrom);
		expect(differsFrom).toBe(vectors.cases[0].sas);
	});

	it("is injective under field-role reassignment (ownerSignPub<->ownerBoxPub swap changes the code)", () => {
		const { admin, enrollee, pin, sas, differsFrom } = vectors.fieldReassign;
		expect(enrollSas(admin, enrollee, pin)).toBe(sas);
		expect(sas).not.toBe(differsFrom);
		expect(differsFrom).toBe(vectors.cases[0].sas);
	});

	it("changes when a committed enrollee key is substituted (the residual MITM detector)", () => {
		const { admin, enrollee, pin, sas, differsFrom } = vectors.substitution;
		expect(enrollSas(admin, enrollee, pin)).toBe(sas);
		expect(sas).not.toBe(differsFrom);
		expect(differsFrom).toBe(vectors.cases[0].sas);
	});

	it("changes when only the pin differs (the OOB shared secret binds the code)", () => {
		const { admin, enrollee, pin } = vectors.cases[0];
		expect(enrollSas(admin, enrollee, `${pin}-x`)).not.toBe(enrollSas(admin, enrollee, pin));
	});

	it("changes when only a domainId differs (Domain binding, closes the confirm->sign edge gap)", () => {
		const { admin, enrollee, pin } = vectors.cases[0];
		const relabeled = { ...enrollee, domainId: `${enrollee.domainId}-x` };
		expect(enrollSas(admin, relabeled, pin)).not.toBe(enrollSas(admin, enrollee, pin));
	});

	it("builds exactly the documented ENROLL_SAS_V1 fixed-slot preimage", () => {
		const { admin, enrollee, pin, sasPreimage, sas } = vectors.handComputed;
		expect(enrollSasPreimage(admin, enrollee, pin).toString("utf8")).toBe(sasPreimage);
		// The documented derivation recomputes the digits independently of the helper:
		// first 8 digest bytes as a big-endian BigInt mod 10^6, zero-padded to 6.
		const digest = crypto.createHash("sha256").update(Buffer.from(sasPreimage, "utf8")).digest();
		let n = 0n;
		for (let i = 0; i < 8; i++) n = (n << 8n) | BigInt(digest[i]);
		const expected = (n % 1_000_000n).toString(10).padStart(6, "0");
		expect(enrollSas(admin, enrollee, pin)).toBe(expected);
		expect(enrollSas(admin, enrollee, pin)).toBe(sas);
	});
});

describe("enroll commitment vectors", () => {
	it("reproduces the committed hash over a party + role + salt", () => {
		const { party, role, salt, commitment } = vectors.commitment;
		expect(enrollCommitment(party, role, salt)).toBe(commitment);
		expect(verifyEnrollCommitment(commitment, party, role, salt)).toBe(true);
	});

	it("the role is load-bearing (the same party+salt under a different role yields a different commitment)", () => {
		const { party, role, salt, commitment, differsFrom } = vectors.commitmentRoleBinding;
		expect(enrollCommitment(party, role, salt)).toBe(commitment);
		expect(commitment).not.toBe(differsFrom);
		expect(differsFrom).toBe(vectors.commitment.commitment);
	});

	it("the salt is load-bearing (a different salt yields a different commitment)", () => {
		const { party, role, salt, commitment, differsFrom } = vectors.commitmentSaltBinding;
		expect(enrollCommitment(party, role, salt)).toBe(commitment);
		expect(commitment).not.toBe(differsFrom);
		expect(differsFrom).toBe(vectors.commitment.commitment);
	});

	it("verifyEnrollCommitment rejects a substituted key, a wrong role, and a wrong salt", () => {
		const { party, role, salt, commitment } = vectors.commitment;
		expect(
			verifyEnrollCommitment(commitment, { ...party, ownerBoxPub: `${party.ownerBoxPub}-x` }, role, salt),
		).toBe(false);
		expect(verifyEnrollCommitment(commitment, party, "ENROLLEE", salt)).toBe(false);
		expect(verifyEnrollCommitment(commitment, party, role, `${salt}-wrong`)).toBe(false);
	});

	it("builds exactly the documented ENROLL_COMMIT_V1 preimage (role-tagged fixed order)", () => {
		const { admin, role, salt, commitmentPreimage, commitment } = vectors.handComputed;
		expect(enrollCommitmentPreimage(admin, role, salt).toString("utf8")).toBe(commitmentPreimage);
		const digest = crypto.createHash("sha256").update(Buffer.from(commitmentPreimage, "utf8")).digest("base64");
		expect(digest).toBe(commitment);
		expect(enrollCommitment(admin, role, salt)).toBe(commitment);
	});
});

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/EnrollSasTest.kt
package com.atelier_nyaarium.switchboard

import com.atelier_nyaarium.switchboard.proto.EnrollParty
import com.atelier_nyaarium.switchboard.proto.SasCrypto
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Drives the hand-authored enroll* SasCrypto twins through the same vectors the vitest
 * suite reads (tests/fixtures/enroll-sas/vectors.json), so the owner-anchored role-tagged
 * enroll SAS cannot drift from the TS source. Unlike the gateway SAS, the preimage is
 * role-FIXED: swapping roles or reassigning fields must change the code.
 */
class EnrollSasTest {
	private val json = Json { ignoreUnknownKeys = true }

	private fun vectors() =
		json.parseToJsonElement(
			javaClass.classLoader!!.getResourceAsStream("enroll-sas/vectors.json")!!.bufferedReader().readText(),
		).jsonObject

	private fun party(o: JsonObject) =
		EnrollParty(
			ownerSignPub = o["ownerSignPub"]!!.jsonPrimitive.content,
			ownerBoxPub = o["ownerBoxPub"]!!.jsonPrimitive.content,
			domainId = o["domainId"]!!.jsonPrimitive.content,
		)

	private fun sasOf(o: JsonObject) =
		SasCrypto.enrollSas(party(o["admin"]!!.jsonObject), party(o["enrollee"]!!.jsonObject), o["pin"]!!.jsonPrimitive.content)

	@Test
	fun sasVectors() {
		for (v in vectors()["cases"]!!.jsonArray) {
			val o = v.jsonObject
			assertEquals(o["sas"]!!.jsonPrimitive.content, sasOf(o))
		}
	}

	@Test
	fun roleFixedAndInjective() {
		// Swapping ADMIN/ENROLLEE roles AND reassigning fields each yield a different code.
		for (key in listOf("roleSwap", "fieldReassign", "substitution")) {
			val o = vectors()[key]!!.jsonObject
			val sas = sasOf(o)
			assertEquals(o["sas"]!!.jsonPrimitive.content, sas)
			assertNotEquals(o["differsFrom"]!!.jsonPrimitive.content, sas)
		}
	}

	@Test
	fun commitmentVector() {
		val c = vectors()["commitment"]!!.jsonObject
		val p = party(c["party"]!!.jsonObject)
		val role = c["role"]!!.jsonPrimitive.content
		val salt = c["salt"]!!.jsonPrimitive.content
		val expected = c["commitment"]!!.jsonPrimitive.content
		assertEquals(expected, SasCrypto.enrollCommitment(p, role, salt))
		assertTrue(SasCrypto.verifyEnrollCommitment(expected, p, role, salt))
	}

	@Test
	fun commitmentRoleBinding() {
		val c = vectors()["commitmentRoleBinding"]!!.jsonObject
		val p = party(c["party"]!!.jsonObject)
		val role = c["role"]!!.jsonPrimitive.content
		val salt = c["salt"]!!.jsonPrimitive.content
		val got = SasCrypto.enrollCommitment(p, role, salt)
		assertEquals(c["commitment"]!!.jsonPrimitive.content, got)
		assertNotEquals(c["differsFrom"]!!.jsonPrimitive.content, got)
	}

	@Test
	fun commitmentSaltBinding() {
		val c = vectors()["commitmentSaltBinding"]!!.jsonObject
		val p = party(c["party"]!!.jsonObject)
		val role = c["role"]!!.jsonPrimitive.content
		val salt = c["salt"]!!.jsonPrimitive.content
		val got = SasCrypto.enrollCommitment(p, role, salt)
		assertEquals(c["commitment"]!!.jsonPrimitive.content, got)
		assertNotEquals(c["differsFrom"]!!.jsonPrimitive.content, got)
		assertFalse(SasCrypto.verifyEnrollCommitment(c["differsFrom"]!!.jsonPrimitive.content, p, role, salt))
	}

	@Test
	fun handComputedPreimages() {
		val hc = vectors()["handComputed"]!!.jsonObject
		val admin = party(hc["admin"]!!.jsonObject)
		val enrollee = party(hc["enrollee"]!!.jsonObject)
		val pin = hc["pin"]!!.jsonPrimitive.content
		val role = hc["role"]!!.jsonPrimitive.content
		val salt = hc["salt"]!!.jsonPrimitive.content

		assertEquals(
			hc["sasPreimage"]!!.jsonPrimitive.content,
			SasCrypto.enrollSasPreimage(admin, enrollee, pin).toString(Charsets.UTF_8),
		)
		assertEquals(
			hc["commitmentPreimage"]!!.jsonPrimitive.content,
			SasCrypto.enrollCommitmentPreimage(admin, role, salt).toString(Charsets.UTF_8),
		)
		assertEquals(hc["sas"]!!.jsonPrimitive.content, SasCrypto.enrollSas(admin, enrollee, pin))
		assertEquals(hc["commitment"]!!.jsonPrimitive.content, SasCrypto.enrollCommitment(admin, role, salt))
	}
}
